### ISSUE TITLE
[Question] Implementation of Optimistic list set (HW5-Optional)

### ISSUE THREAD

**Issue by ICubE-** [kaist-cp/cs431#902]
The comment says:
```
/// An iterator visiting all elements. `next()` returns `Some(Err(()))` when validation fails.
/// In that case, the user must restart the iteration.
```
1. What is 'validation'? Is it the same condition as returning `Err(())` in `Cursor::find`?
Also, if `Some(Err(())` is returned, then does the `Iter` object become unusable? What happens if `Iter::next` is called again?
I'm not sure where should I manually drop the `cursor`.
2. How do I clean up `cursor.prev` after the use of `Iter`? 
Since `ReadGuard` must not be dropped, it has to be cleaned up using `ReadGuard::finish` before `Iter` is destroyed. But the function consumes the ownership, so I cannot call it in `Iter::next`.

**Comment by Lee-Janggun** [comment]
1. Yes, it's the same. Once 'Some(Err())' is returned, the next calls to 'next()' should also return 'Some(Err())'.
2. You don't need to worry about accidentally dropping it; the ManuallyDrop prevents that.

**Comment by ICubE-** [comment]
But I have to eventually drop the `cursor` to free its allocated memory. I have no idea with 'how' and 'when'.
Is there a way to take ownership of `cursor.prev` inside `next()`? When should I destroy `cursor`?
Or do I have to change the structure of `Iter`?

**Comment by Lee-Janggun** [comment]
A cursor has allocated memory associated with it; it has just a `ReadGuard` and a `Shared`. Just ending it's lifetime (which is done automatically when the outer `Iter`'s lifetime ends) is sufficient. In other words, you don't need to worry about it.

**Comment by ICubE-** [comment]
I understood that I don't need to worry about ManuallyDrop.

Now I have other questions:
3. `Cursor::find` returns `Err(())` if the cursor cannot move. Is it the situation that prev and curr are not consistent? I think cursors can always move if curr is not null, regardless of prev.
4. If `ReadGuard::upgrade` fails, then do I have to run `find` again?

**Comment by Lee-Janggun** [comment]
3. Yes it isn't consistent. I.e., the next node of `prev` isn't `curr`. This may seem ok, you can try do it. But it can actually cause an double-free. The exact details are for you to figure out. (Hint: consider what will happen if you just do traversals in inconsistent situations, especially a long chain of already removed nodes.)
4. Yes you should restart from the beginning.

**Comment by ICubE-** [comment]
Now I'm struggling in the situation that `Cursor::find` is waiting for a read lock of a node that has been already removed.

The doc says like following, but I don't get it;
> Since a node can be removed while another thread is reading, reclamation of the node should be deferred. You can handle this semi-automatically with `crossbeam_epoch`.

Is there an extra hint about this quote?

**Comment by Lee-Janggun** [comment]
It means you should use 'Guard::defer_destroy'.

**Comment by ICubE-** [comment]
Thank you for your kind answer!

I have one last question about `Iter::next`.
Consider the situation calling `Iter::next`, and its cursor is consistent at that point. Then the function will save the current data to return, and progress prev and curr to the next one. Just before returning the function, if the progressed prev - which is prev of next `Iter::next` - becomes invalid, does the function is allowed to return `Some(Err(()))`?
In other words, does 'validation' check only the consistency of this time, not including next time beforehand?

**Comment by Lee-Janggun** [comment]
No, you should only check the consistency of one cursor state. Don't check the consistency of future next calls
